### PR TITLE
ELASTICSEARCH_ENCRYPTED_AT_REST - Pushing newly created rule

### PR DESCRIPTION
**Pull request by ShrikantDavange**
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

*Issue #, if available:*

*Description of changes:*

### CODE REVIEW

**Comment by jongogogo** [comment]
You got to clean up the errors?

**Comment by ShrikantDavange** [comment]
I pushed changes in ELASTICSEARCH_ENCRYPTED_AT_REST and made pylint score to 10. The git hook is taking the pylint of ELASTICSEARCH_IN_VPC_ONLY.

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    es_client = get_client('es', event)
+    list_domain = es_client.list_domain_names()
+    evaluations = []
+    if not list_domain['DomainNames']:
+        return 'NOT_APPLICABLE'
+    domain_list = []
+    for domain in list_domain['DomainNames']:
+        domain_list.append(domain['DomainName'])
+
+    domain_config = es_client.describe_elasticsearch_domains(DomainNames=domain_list)
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: You need to get all domains:
Check https://github.com/awslabs/aws-config-rules/blob/master/python/ELASTICSEARCH_IN_VPC_ONLY/ELASTICSEARCH_IN_VPC_ONLY.py
with the function `def get_all_domain_details(es_client, es_domains):`

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    for domain in domain_config['DomainStatusList']:
+        if domain['EncryptionAtRestOptions']['Enabled']:
+            evaluations.append(build_evaluation(domain['DomainName'], 'COMPLIANT', event))
+        else:
+            evaluations.append(build_evaluation(domain['DomainName'], 'NON_COMPLIANT', event, annotation='The Amazon Elasticsearch domain does not have the encryption of data at rest as enabled'))
+    return evaluations
+
+
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: remove extra line

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    domain_list = []
+    for domain in list_domain['DomainNames']:
+        domain_list.append(domain['DomainName'])
+
+    domain_config = es_client.describe_elasticsearch_domains(DomainNames=domain_list)
+    for domain in domain_config['DomainStatusList']:
+        if domain['EncryptionAtRestOptions']['Enabled']:
+            evaluations.append(build_evaluation(domain['DomainName'], 'COMPLIANT', event))
+        else:
+            evaluations.append(build_evaluation(domain['DomainName'], 'NON_COMPLIANT', event, annotation='The Amazon Elasticsearch domain does not have the encryption of data at rest as enabled'))
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: Correct to `This Amazon Elasticsearch domain does not have encryption of data at rest enabled.`

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    es_client = get_client('es', event)
+    list_domain = es_client.list_domain_names()
+    evaluations = []
+    if not list_domain['DomainNames']:
+        return 'NOT_APPLICABLE'
+    domain_list = []
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: Can be replaced by my comment below.

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
+import sys
+import unittest
+try:
+    from unittest.mock import MagicMock
+except ImportError:
+    from mock import MagicMock
+import botocore
+
+DEFAULT_RESOURCE_TYPE = 'AWS::Elasticsearch::Domain'
+CONFIG_CLIENT_MOCK = MagicMock()
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: Need extraline

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            return STS_CLIENT_MOCK
+        if client_name == 'es':
+            return ES_CLIENT_MOCK
+        raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('ELASTICSEARCH_ENCRYPTED_AT_REST')
+
+class SampleTest(unittest.TestCase):
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: Change name to ComplianceTest

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    def test_scenario_1_is_null_domains(self):
+        ES_CLIENT_MOCK.list_domain_names = MagicMock(return_value={'DomainNames': []})
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NOT_APPLICABLE', '123456789012', 'AWS::Elasticsearch::Domain'))
+        assert_successful_evaluation(self, response, resp_expected, 1)
+
+    def test_scenario_1_is_compliant(self):
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: scenario_2 ?

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        describe_domain = {"DomainStatusList": [{"EncryptionAtRestOptions":{"Enabled":True}, "DomainName":"test"}]}
+        ES_CLIENT_MOCK.describe_elasticsearch_domains = MagicMock(return_value=describe_domain)
+
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('COMPLIANT', 'test', 'AWS::Elasticsearch::Domain'))
+        assert_successful_evaluation(self, response, resp_expected, 1)
+
+    def test_scenario_1_is_non_compliant(self):
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: scenario_3 ?

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    def test_scenario_1_is_null_domains(self):
+        ES_CLIENT_MOCK.list_domain_names = MagicMock(return_value={'DomainNames': []})
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NOT_APPLICABLE', '123456789012', 'AWS::Elasticsearch::Domain'))
+        assert_successful_evaluation(self, response, resp_expected, 1)
+
+    def test_scenario_1_is_compliant(self):
+        list_domains = {'DomainNames':[{'DomainName':'test'}]}
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: Test with at least 2 domains.

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        ES_CLIENT_MOCK.describe_elasticsearch_domains = MagicMock(return_value=describe_domain)
+
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('COMPLIANT', 'test', 'AWS::Elasticsearch::Domain'))
+        assert_successful_evaluation(self, response, resp_expected, 1)
+
+    def test_scenario_1_is_non_compliant(self):
+        list_domains = {'DomainNames': [{'DomainName': 'test'}]}
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: Test with at least 2 domains

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+    if not es_domain_list:
+        return 'NOT_APPLICABLE'
+
+    es_domain_list_details = get_all_domain_details(es_client, es_domain_list)
+
+    for es_domain_details in es_domain_list_details:
+        if es_domain_details['EncryptionAtRestOptions']['Enabled']:
+            evaluations.append(build_evaluation(es_domain_details['DomainName'], 'COMPLIANT', event))
+        else:
+            evaluations.append(build_evaluation(es_domain_details['DomainName'], 'NON_COMPLIANT', event, annotation='This Amazon Elasticsearch domain does not have the encryption of data at rest as enabled'))
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: `This Amazon Elasticsearch domain does not have the encryption of data at rest as enabled`
to
`This Amazon Elasticsearch domain is not encrypted at rest.`

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py (hunk 5 above)
Comment (jongogogo) [reply python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: ?

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    def setUp(self):
+        pass
+
+    def test_scenario_1_is_null_domains(self):
+        ES_CLIENT_MOCK.list_domain_names = MagicMock(return_value={'DomainNames': []})
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('NOT_APPLICABLE', '123456789012', 'AWS::Elasticsearch::Domain'))
+        assert_successful_evaluation(self, response, resp_expected, 1)
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: 1 is default value, no needed here

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+Trigger:
+    Periodic
+Reports on:
+    AWS::Elasticsearch::Domain
+Rule Parameters:
+    None
+
+Scenarios:
+  Scenario: 1
+     Given: No Amazon Elasticsearch Service (Amazon ES) domains are present in a region for an account.
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: Just `Given: No Amazon Elasticsearch Service (Amazon ES) domains are present`

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        domain_details = es_client.describe_elasticsearch_domains(DomainNames=es_domains_names_only[:5])['DomainStatusList']
+        es_domain_list_details += domain_details
+        del es_domains_names_only[:5]
+    return es_domain_list_details
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    es_client = get_client('es', event)
+    es_domain_list = es_client.list_domain_names()['DomainNames']
+    evaluations = []
+    print(configuration_item)
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST.py]: Remove

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py (hunk 5 above)
Comment (jongogogo) [reply python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: ?

File: python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+    def test_scenario_3_is_non_compliant(self):
+        ES_CLIENT_MOCK.list_domain_names = MagicMock(return_value=self.list_domains_scenario_2)
+        ES_CLIENT_MOCK.describe_elasticsearch_domains = MagicMock(return_value=self.describe_domain_scenario_2)
+        lambda_event = build_lambda_scheduled_event(rule_parameters=None)
+        response = RULE.lambda_handler(lambda_event, {})
+        print(response)
+        resp_expected = []
+        resp_expected.append(build_expected_response('NON_COMPLIANT', 'domain1', 'AWS::Elasticsearch::Domain', annotation='This Amazon Elasticsearch domain is not encrypted at rest.'))
+        resp_expected.append(build_expected_response('NON_COMPLIANT', 'domain2', 'AWS::Elasticsearch::Domain', annotation='This Amazon Elasticsearch domain is not encrypted at rest.'))
```
Comment (jongogogo) [inline python/ELASTICSEARCH_ENCRYPTED_AT_REST/ELASTICSEARCH_ENCRYPTED_AT_REST_test.py]: No need to put the default resource. It is passed by default. Review all similar case.

**Pull request merged**